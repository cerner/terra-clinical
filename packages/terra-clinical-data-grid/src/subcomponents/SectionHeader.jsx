import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import IconCaretRight from 'terra-icon/lib/icon/IconCaretRight';
import IconCaretDown from 'terra-icon/lib/icon/IconCaretDown';

import { KEYCODES } from '../utils/keycodes';

import styles from './SectionHeader.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * String identifier of the section that the SectionHeader is rendered within.
   */
  sectionId: PropTypes.string.isRequired,
  /**
   * String text rendered within the SectionHeader's default text position.
   */
  text: PropTypes.string,
  /**
   * Content rendered within the SectionHeader's default leading content position (ahead of the text).
   */
  startAccessory: PropTypes.node,
  /**
   * Content rendered within the SectionHeader's default trailing content position (after the text).
   */
  endAccessory: PropTypes.node,
  /**
   * Boolean indicating whether the SectionHeader is representing a collapsible section. If true, the SectionHeader
   * will be selectable, and an collapsiblity indicator will be rendered within the SectionHeader.
   */
  isCollapsible: PropTypes.bool,
  /**
   * Boolean indicating whether the SectionHeader is representing a collapsed section.
   */
  isCollapsed: PropTypes.bool,
  /**
   * Function called upon selection of the SectionHeader. The `isCollapsible` prop must be true for this function
   * to be executed. Parameters: `onRequestSectionCollapse(sectionId)`
   */
  onRequestSectionCollapse: PropTypes.func,
  /**
   * Function that will be called with a ref to the SectionHeader's selectable element. Parameters: `selectableRefCallback(selectableRef)`
   */
  selectableRefCallback: PropTypes.func,
  /**
   * Content to be rendered within the SectionHeader. If provided, the `text`, `startAccessory`, and `endAccessory` props are ignored.
   */
  children: PropTypes.node,
};

class SectionHeader extends React.Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  handleClick() {
    const { onRequestSectionCollapse, sectionId } = this.props;

    if (onRequestSectionCollapse) {
      onRequestSectionCollapse(sectionId);
    }
  }

  handleKeyDown(event) {
    if (event.nativeEvent.keyCode === KEYCODES.ENTER || event.nativeEvent.keyCode === KEYCODES.SPACE) {
      const { onRequestSectionCollapse, sectionId } = this.props;

      if (onRequestSectionCollapse) {
        event.preventDefault();
        onRequestSectionCollapse(sectionId);
      }
    }
  }

  render() {
    const {
      sectionId, text, startAccessory, endAccessory, children, isCollapsible, isCollapsed, selectableRefCallback,
    } = this.props;

    let content;
    if (children) {
      content = (
        <div className={cx('flex-content')}>
          {children}
        </div>
      );
    } else {
      content = (
        <React.Fragment>
          {startAccessory ? (
            <div className={cx('start-accessory')}>
              {startAccessory}
            </div>
          ) : undefined}
          <div className={cx('flex-content')}>
            {text ? <div className={cx('text')}>{text}</div> : null}
          </div>
          {endAccessory ? (
            <div className={cx('end-accessory')}>
              {endAccessory}
            </div>
          ) : undefined}
        </React.Fragment>
      );
    }

    return (
      <div
        className={cx('section-header')}
        data-section-header
        data-terra-clinical-data-grid-section-header-id={sectionId}
      >
        <div
          role="button"
          className={cx(['content', { selectable: isCollapsible }])}
          onClick={isCollapsible ? this.handleClick : undefined}
          onKeyDown={isCollapsible ? this.handleKeyDown : undefined}
          tabIndex={isCollapsible ? '0' : undefined}
          ref={selectableRefCallback}
        >
          {isCollapsible ? (
            <div className={cx('collapsible-icon')}>
              {isCollapsed ? <IconCaretRight /> : <IconCaretDown />}
            </div>
          ) : null}
          {content}
        </div>
      </div>
    );
  }
}

SectionHeader.propTypes = propTypes;

export default SectionHeader;
