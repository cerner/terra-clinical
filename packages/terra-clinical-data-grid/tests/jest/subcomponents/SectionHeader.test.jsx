import React from 'react';
import SectionHeader from '../../../src/subcomponents/SectionHeader';

describe('SectionHeader Snapshot Tests', () => {
  it('should render a SectionHeader with minimal props', () => {
    const sectionHeader = shallow((
      <SectionHeader sectionId="test-section" />
    ));

    expect(sectionHeader).toMatchSnapshot();
  });

  it('should render a SectionHeader with default text/accessory props', () => {
    const sectionHeader = shallow((
      <SectionHeader
        sectionId="test-section"
        text="Default Text"
        startAccessory={<div>Start</div>}
        endAccessory={<div>End</div>}
        isCollapsible
        onRequestSectionCollapse={() => {}}
        selectableRefCallback={() => {}}
      />
    ));

    expect(sectionHeader).toMatchSnapshot();
  });

  it('should render a SectionHeader with child content if provided', () => {
    const sectionHeader = shallow((
      <SectionHeader
        sectionId="test-section"
        text="Default Text"
        startAccessory={<div>Start</div>}
        endAccessory={<div>End</div>}
        isCollapsible
        onRequestSectionCollapse={() => {}}
        selectableRefCallback={() => {}}
      >
        <div>Custom Section Header Layout</div>
      </SectionHeader>
    ));

    expect(sectionHeader).toMatchSnapshot();
  });

  it('should render a SectionHeader as collapsed', () => {
    const sectionHeader = shallow((
      <SectionHeader
        sectionId="test-section"
        text="Default Text"
        startAccessory={<div>Start</div>}
        endAccessory={<div>End</div>}
        isCollapsible
        isCollapsed
        onRequestSectionCollapse={() => {}}
        selectableRefCallback={() => {}}
      />
    ));

    expect(sectionHeader).toMatchSnapshot();
  });
});
