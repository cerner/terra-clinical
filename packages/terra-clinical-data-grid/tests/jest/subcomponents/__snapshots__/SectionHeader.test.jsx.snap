// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionHeader Snapshot Tests should render a SectionHeader as collapsed 1`] = `
<div
  className="section-header"
  data-section-header={true}
>
  <div
    className="content selectable"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="collapsible-icon"
    >
      <IconCaretRight
        className=""
        isBidi={true}
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
    <div
      className="start-accessory"
    >
      <div>
        Start
      </div>
    </div>
    <div
      className="flex-content"
    >
      <div
        className="text"
      >
        Default Text
      </div>
    </div>
    <div
      className="end-accessory"
    >
      <div>
        End
      </div>
    </div>
  </div>
</div>
`;

exports[`SectionHeader Snapshot Tests should render a SectionHeader with child content if provided 1`] = `
<div
  className="section-header"
  data-section-header={true}
>
  <div
    className="content selectable"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="collapsible-icon"
    >
      <IconCaretDown
        className=""
        isBidi={true}
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
    <div
      className="flex-content"
    >
      <div>
        Custom Section Header Layout
      </div>
    </div>
  </div>
</div>
`;

exports[`SectionHeader Snapshot Tests should render a SectionHeader with default text/accessory props 1`] = `
<div
  className="section-header"
  data-section-header={true}
>
  <div
    className="content selectable"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="collapsible-icon"
    >
      <IconCaretDown
        className=""
        isBidi={true}
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
    <div
      className="start-accessory"
    >
      <div>
        Start
      </div>
    </div>
    <div
      className="flex-content"
    >
      <div
        className="text"
      >
        Default Text
      </div>
    </div>
    <div
      className="end-accessory"
    >
      <div>
        End
      </div>
    </div>
  </div>
</div>
`;

exports[`SectionHeader Snapshot Tests should render a SectionHeader with minimal props 1`] = `
<div
  className="section-header"
  data-section-header={true}
>
  <div
    className="content"
    role="button"
  >
    <div
      className="flex-content"
    />
  </div>
</div>
`;
